src/opt/ReplaceHalfMaskedMemOps.cpp

```diff
@@ -0,0 +1,216 @@
+/*
+  Copyright (c) 2024, Intel Corporation
+
+  SPDX-License-Identifier: BSD-3-Clause
+*/
+
+#include "ReplaceHalfMaskedMemOps.h"
+
+#include <llvm/IR/IRBuilder.h>
+
+namespace ispc {
+
+bool lIsSecondHalfAllFalse(llvm::Value *mask) {
+    if (auto *CV = llvm::dyn_cast<llvm::ConstantVector>(mask)) {
+        auto N = CV->getType()->getNumElements();
+        for (auto i = N / 2; i < N; i++) {
+            llvm::Constant *E = CV->getAggregateElement(i);
+            if (!E || !llvm::isa<llvm::ConstantInt>(E) || !llvm::cast<llvm::ConstantInt>(E)->isZero()) {
+                // found non-zero element
+                return false;
+            }
+        }
+        for (auto i = 0; i < N / 2; i++) {
+            llvm::Constant *E = CV->getAggregateElement(i);
+            if (!E || !llvm::isa<llvm::ConstantInt>(E) || !llvm::cast<llvm::ConstantInt>(E)->isOne()) {
+                // found non-one element
+                return false;
+            }
+        }
+        return true;
+    }
+    return false;
+}
+
+llvm::Value *lMergeVectors(llvm::IRBuilder<> &B, llvm::Value *firstVector, llvm::Value *secondVector,
+                           llvm::Twine &name) {
+    auto *firstVecType = llvm::cast<llvm::VectorType>(firstVector->getType());
```

Comment:
Why do you use `llvm::cast` and not `llvm::dyn_cast`?

```
auto *firstVecType = llvm::dyn_cast<llvm::VectorType>(firstVector->getType());
auto *secondVecType = llvm::dyn_cast<llvm::VectorType>(secondVector->getType());
```

Reply:
Done

Reply:
I still see mix of `llvm::dyn_cast` and `llvm::cast` in the code. It's not a problem if you're certain of the type you are casting to, and the program logic guarantees that the type is correct. If it's not, I suggest using `llvm::dyn_cast`.

Reply:
I have changed all `llvm::cast` to `llvm::dyn_cast`